providers/gitlab.go

```diff
@@ -250,9 +314,40 @@ func (p *GitLabProvider) GetEmailAddress(ctx context.Context, s *sessions.Sessio
 		return "", fmt.Errorf("group membership check failed: %v", err)
 	}
 
+	// Check project membership
+	err = p.verifyProjectMembership(ctx, s)
+	if err != nil {
+		return "", fmt.Errorf("project membership check failed: %v", err)
+	}
+
 	return userInfo.Email, nil
 }
 
+func (p *GitLabProvider) verifyProjectMembership(ctx context.Context, s *sessions.SessionState) error {
+	if len(p.Projects) == 0 {
+		return nil
+	}
+
+	for _, project := range p.Projects {
+		project, err := p.getProjectInfo(ctx, s, project)
+		if err == nil && !project.Archived {
```

Comment:
I'm assuming no error means that the user is a member of the project? This could probably do with a comment to explain if that's the case